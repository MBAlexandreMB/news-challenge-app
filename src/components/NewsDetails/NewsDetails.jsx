import Cross from '../../shared/icons/Cross';
import { ICON_SIZES, NO_IMAGE_URL } from '../../shared/contants';
import { onImageLoadingError } from '../../shared/utils';
import './NewsDetails.scss';

const NewsDetails = ({ news, onClose }) => {
    const { title, description, url, content } = news;
    const urlToImage = news.urlToImage || NO_IMAGE_URL;

    let contentText = null;
    if (content) {
        const contentEndIdx = content.lastIndexOf('â€¦');
        contentText = contentEndIdx !== -1
            ? content.substring(0, contentEndIdx + 1)
            : content;
    }

    return (
        <article className='news-details-container'>
            <figure className='news-image'>
                <button className='close-details-button' onClick={onClose}>
                    <Cross size={ICON_SIZES.S} />
                </button>
                <img
                    src={urlToImage}
                    loading='lazy'
                    onError={onImageLoadingError}
                    alt='' // The API returns no description of the image
                />
            </figure>
            <div className='news-title-content-container'>
                <h1>{title}</h1>
                <h2>{description}</h2>
                <p  className='content'>
                    <span data-test="content">{contentText}</span>
                    <a href={url} target="_blank" rel='noreferrer'>[continue reading]</a>
                </p>
            </div>
        </article>
    );
}
 
export default NewsDetails;