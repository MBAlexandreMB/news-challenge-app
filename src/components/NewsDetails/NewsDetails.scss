.news-details-container {
    background-color: #fff;
    border-radius: 4px;
    padding: 10px 15px 20px;
    position: fixed;
    right: 50px;
    top: 20px;
    width: 50vw;

    .news-title-image-container {
        display: flex;
        flex-flow: row nowrap;

        h4 {
            margin-top: 6px;
            font-weight: 200;
        }

        .news-image {
            min-width: 200px;
            max-width: 200px;
            width: 200px;
            height: 200px;
            margin: 4px 15px 0 0;
            
            img {
                border-radius: 4px;
                width: 100%;
                height: 100%;
                object-fit: cover;
            }
        }
    }
    
}