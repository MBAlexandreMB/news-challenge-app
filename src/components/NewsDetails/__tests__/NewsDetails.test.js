import { shallow } from 'enzyme';
import NewsDetails from '../NewsDetails';

const defaultProps = {
    news: {
        title: 'fakeTitle',
        description: 'fakeDescription',
        url: 'fakeUrl',
        content: 'fakeContent',
    },
    onClose: jest.fn(),
};

describe('<NewsListing />', () => {
    it('should render correctly', () => {
        const wrapper = shallow(<NewsDetails {...defaultProps} />);

        expect(wrapper).toHaveLength(1);
        expect(wrapper).toMatchSnapshot();
    });

    it('should render correctly with a news with no values', () => {
        const wrapper = shallow(<NewsDetails {...defaultProps} news={{}} />);

        expect(wrapper).toHaveLength(1);
        expect(wrapper).toMatchSnapshot();
    });

    it('should remove anything after "…" on content', () => {
        const updatedNews = {
            ...defaultProps.news,
            content: 'a content with… [and other characteres]'
        };
        const wrapper = shallow(<NewsDetails {...defaultProps} news={updatedNews} />);
        const content = wrapper.find('[data-test="content"]').text();

        expect(content).toBe('a content with…');
    });
});