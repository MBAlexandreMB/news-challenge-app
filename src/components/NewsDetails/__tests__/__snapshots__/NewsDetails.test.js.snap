// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NewsListing /> should render correctly 1`] = `
<article
  className="news-details-container"
>
  <figure
    className="news-image"
  >
    <button
      className="close-details-button"
      onClick={[MockFunction]}
    >
      <Cross
        size={10}
      />
    </button>
    <img
      alt=""
      loading="lazy"
      onError={[Function]}
      src="https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcRBJvWZNK1Lm7z-Iw0g4RJvw6l1sPwOUDOaZW_AecE3ZZdPyy4KpWejikv7VSD4cuUQpOY&usqp=CAU"
    />
  </figure>
  <div
    className="news-title-content-container"
  >
    <h1>
      fakeTitle
    </h1>
    <h2>
      fakeDescription
    </h2>
    <p
      className="content"
    >
      <span
        data-test="content"
      >
        fakeContent
      </span>
      <a
        href="fakeUrl"
        rel="noreferrer"
        target="_blank"
      >
        [continue reading]
      </a>
    </p>
  </div>
</article>
`;

exports[`<NewsListing /> should render correctly with a news with no values 1`] = `
<article
  className="news-details-container"
>
  <figure
    className="news-image"
  >
    <button
      className="close-details-button"
      onClick={[MockFunction]}
    >
      <Cross
        size={10}
      />
    </button>
    <img
      alt=""
      loading="lazy"
      onError={[Function]}
      src="https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcRBJvWZNK1Lm7z-Iw0g4RJvw6l1sPwOUDOaZW_AecE3ZZdPyy4KpWejikv7VSD4cuUQpOY&usqp=CAU"
    />
  </figure>
  <div
    className="news-title-content-container"
  >
    <h1 />
    <h2 />
    <p
      className="content"
    >
      <span
        data-test="content"
      />
      <a
        rel="noreferrer"
        target="_blank"
      >
        [continue reading]
      </a>
    </p>
  </div>
</article>
`;
